internal/auth/oidc/provider.go

```diff
@@ -0,0 +1,220 @@
+			u.Host = addr.Addr
+			if !addr.Tls {
+				u.Scheme = "http"
+			}
+			u.Path = "/oidc/callback"
+
+			allowedUris = append(allowedUris, u.String())
+		}
+	}
+
+	return oidc.NewConfig(
+		am.DiscoveryUrl,
+		am.ClientId,
+		oidc.ClientSecret(am.ClientSecret),
+		algs,
+		allowedUris,
+		oidc.WithAudiences(am.Auds...),
+		oidc.WithProviderCA(strings.Join(am.DiscoveryCaPem, "\n")),
+	)
+}
+
+// ProviderCache is a cache for OIDC providers. OIDC providers are something
+// you don't want to recreate per-request since they make HTTP requests
+// themselves.
+//
+// The ProviderCache purges a provider under two scenarios: (1) the
+// provider config is updated and it is different and (2) after a set
+// amount of time (see cacheExpiry for value) in case the remote provider configuration
+// changed.
+type ProviderCache struct {
+	providers map[string]*oidc.Provider
+	mu        sync.RWMutex
+	cancel    context.CancelFunc
+}
+
+// NewProviderCache should be used to initialize a provider cache. This
+// will start up background resources to manage the cache.
+func NewProviderCache() *ProviderCache {
+	ctx, cancel := context.WithCancel(context.Background())
+	result := &ProviderCache{
+		providers: map[string]*oidc.Provider{},
+		cancel:    cancel,
+	}
+
+	// Start the cleanup timer
+	go result.runCleanupLoop(ctx)
+
+	return result
+}
+
+// Get returns the OIDC provider for the given auth method configuration.
+// This will initialize the provider if it isn't already in the cache or
+// if the configuration changed.
+func (c *ProviderCache) Get(
+	ctx context.Context, am *pb.AuthMethod, sc *pb.ServerConfig,
+) (*oidc.Provider, error) {
+	amMethod, ok := am.Method.(*pb.AuthMethod_Oidc)
+	if !ok {
+		return nil, errors.New("auth method must be OIDC")
+	}
+
+	// No matter what we'll use the config of the arg method since we'll
+	// use it to compare to existing (if exists) or initialize a new provider.
+	oidcCfg, err := ProviderConfig(amMethod.Oidc, sc)
+	if err != nil {
+		return nil, err
+	}
+
+	// Normalize name
+	name := strings.ToLower(am.Name)
+
+	// Get our current value
+	var current *oidc.Provider
+	ok = false
+	c.mu.RLock()
+	if c.providers != nil {
+		current, ok = c.providers[name]
+	}
+	c.mu.RUnlock()
+
+	// If we have a current value, we want to compare hashes to detect changes.
+	if ok {
+		currentHash, err := current.ConfigHash()
+		if err != nil {
+			return nil, err
+		}
+
+		newHash, err := oidcCfg.Hash()
+		if err != nil {
+			return nil, err
+		}
+
+		// If the hashes match, this is cached.
+		if currentHash == newHash {
+			return current, nil
+		}
+	}
+
+	// If we made it here, the provider isn't in the cache OR the config changed.
+	// Initialize a new provider.
+	newProvider, err := oidc.NewProvider(oidcCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// If we have an old provider, clean up resources.
+	if current != nil {
+		current.Done()
+	}
+
+	c.providers[name] = newProvider
```

Comment:
I seem to understand that `c.providers` might be `nil` here due to a previous `Clear()` and thus the code might panic at this line

Reply:
Good catch. Fixed!